plugins/apps/threatbus_zeek/plugin.py

```diff
@@ -11,129 +12,180 @@
 )
-import time
+from typing import Callable, Dict, List, Union
 
 """Zeek network monitor - plugin for Threat Bus"""
 
 plugin_name = "zeek"
 lock = threading.Lock()
-subscriptions = dict()
-
-
-def validate_config(config):
-    assert config, "config must not be None"
-    config["host"].get(str)
-    config["port"].get(int)
-    config["module_namespace"].get(str)
-
-
-def rand_string(length):
-    """Generates a pseudo-random string with the requested length"""
-    letters = string.ascii_lowercase
-    return "".join(random.choice(letters) for i in range(length))
-
-
-def manage_subscription(
-    ep, module_namespace, task, subscribe_callback, unsubscribe_callback
-):
-    global lock, subscriptions
-    rand_suffix_length = 10
-    if type(task) is Subscription:
-        # point-to-point topic and queue for that particular subscription
-        logger.info(f"Received subscription for topic: {task.topic}")
-        p2p_topic = task.topic + rand_string(rand_suffix_length)
-        p2p_q = Queue()
-        ack = broker.zeek.Event(
-            f"{module_namespace}::subscription_acknowledged", p2p_topic
-        )
-        ep.publish("threatbus/manage", ack)
-        subscribe_callback(task.topic, p2p_q, task.snapshot)
-        lock.acquire()
-        subscriptions[p2p_topic] = p2p_q
-        lock.release()
-    elif type(task) is Unsubscription:
-        logger.info(f"Received unsubscription from topic: {task.topic}")
-        threatbus_topic = task.topic[: len(task.topic) - rand_suffix_length]
-        p2p_q = subscriptions.get(task.topic, None)
-        if p2p_q:
-            unsubscribe_callback(threatbus_topic, p2p_q)
+subscriptions: Dict[str, JoinableQueue] = dict()  # p2p_topic => queue
+workers: List[threatbus.StoppableWorker] = list()
+
+
+class SubscriptionManager(threatbus.StoppableWorker):
+    def __init__(
+        self,
+        module_namespace: str,
+        ep: broker.Endpoint,
+        subscribe_callback: Callable,
+        unsubscribe_callback: Callable,
+    ):
+        """
+        @param module_namespace A Zeek namespace to accept events from
+        @param ep The broker endpoint used for listening
+        @param subscribe_callback The callback to invoke for new subscriptions
+        @param unsubscribe_callback The callback to invoke for revoked subscriptions
+        """
+        super(SubscriptionManager, self).__init__()
+        self.ep = ep
+        self.module_namespace = module_namespace
+        self.subscribe_callback = subscribe_callback
+        self.unsubscribe_callback = unsubscribe_callback
+        self.rand_suffix_length = 10
+
+    def run(self):
+        """
+        Binds a broker subscriber to the given endpoint. Only listens for management
+        messages, such as un/subscriptions of new clients.
+        """
+        global logger
+        sub = self.ep.make_subscriber("threatbus/manage")
+        while self._running():
+            (ready_readers, [], []) = select.select([sub.fd()], [], [], 1)
```

Comment:
I was wondering: why not `select.epoll`? Portability? If you support `epoll` then you'd also have to do `kqueue`...and then maybe you want something like `libuv`.

Reply:
Yes, portability. From [the docs](https://docs.python.org/3/library/select.html#select.epoll): Epoll is "Only supported on Linux 2.5.44 and newer."

So folks using a Mac could not develop on Threat Bus.

Reply:
Okay, I figured. This will work for a long time until it the actual bottleneck.